Fix endianness issue in Tensor::BitcastFrom() on s390x

kun-lu20 opened the pull request:
The following Bitcast related test cases would fail on s390x (BE machines):

- `//tensorflow/python/ops/ragged:ragged_bitcast_op_test`
- `//tensorflow/compiler/tests:cast_ops_test_cpu`
- `//tensorflow/compiler/tests:cast_ops_test_cpu_mlir_bridge_test`

I found the root cause is that the basic Bitcast operation in `Tensor::BitcastFrom()` only reuses the input buffer, see https://github.com/tensorflow/tensorflow/blob/55782dc6b8276eddce25f56fe97aacd0fc3c6fe7/tensorflow/core/framework/tensor.cc#L757, which would cause the endianness issue on BE machines when bit-casting between types of different sizes.

This PR made code changes in `tensor.cc` which use `tensor::DeepCopy()` function to allocate a new buffer for the output in `BitcastFrom()` function on BE machines when the types are of different sizes, since the output buffer needs to be swapped to solve the endianness issue and the input buffer shouldn’t be affected. After applying this PR, the above test cases could pass on BE machines.

This code change would not cause any regressions on existing test cases and it won't affect LE machines.

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 committed:
Fix several Bitcast related test cases on s390x

Fix the following test cases on s390x:
//tensorflow/python/ops/ragged:ragged_bitcast_op_test
//tensorflow/compiler/tests:cast_ops_test_cpu
//tensorflow/compiler/tests:cast_ops_test_cpu_mlir_bridge_test

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 committed:
Fix object scope issue for ts_.buf_

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Hi @sagunb @cantonios ,

Could you please kindly review this PR? Thank you very much!

sagunb commented:
Hi Kun-Lu thanks for the fix! Just a quick comment - the official bitcast doc https://www.tensorflow.org/api_docs/python/tf/bitcast mentions that there is no data copying so the buffer re-use is a critical requirement. If we wanted to break backwards compatibility then there would have to be an RFC process for it. Is there a need to create a copy of the in buffer for the out buffer - can we not do the transformations on the out buffer?

kun-lu20 commented:
Hi @sagunb ,

Thanks so much for your quick reply.

If we skip the data swapping on the outer buffer, the result on BE platforms will differ from LE platforms due to different endian orderings.

If we reuse the input buffer and keep the data swapping, the content of input buffer will be changed as well. However, we can make a slight adjustment in `cast_ops_test.py` (as shown in the following patch), so that all the above test cases could still pass.
```diff
diff --git a/tensorflow/compiler/tests/cast_ops_test.py b/tensorflow/compiler/tests/cast_ops_test.py
index 3c11057d930..f3d61b4ef29 100644
--- a/tensorflow/compiler/tests/cast_ops_test.py
+++ b/tensorflow/compiler/tests/cast_ops_test.py
@@ -40,8 +40,8 @@ class CastOpsTest(xla_test.XLATestCase):
 
       x = random_ops.random_normal([10, 10, 2], dtype=dtypes.float16)
       with ops.device(self.device):
-        out = f(x)
         compiled_out = compiled_f(x)
+        out = f(x)
         self.assertAllClose(out, compiled_out)
         # 10,10,2--(bitcast-convert)-->10,10--(reduce)-->10
         self.assertEqual(out.shape[0], 10)
```

I've also noticed the following note from the documentation:
> Note: Bitcast is implemented as a low-level cast, so machines with different endian orderings will give different results. 

Although the documentation has mentioned that, I think it would be better to make BE machines get the same results as LE machines in these test cases.

For example, in test case `//tensorflow/compiler/tests:cast_ops_test_cpu`, I think it's comparing the results between the JIT Compiled XLA Bitcast operation with the basic tensor Bitcast operation, as shown in the below code:
https://github.com/tensorflow/tensorflow/blob/6cd264ba65e6161164d7f1b463b7d2415ac0aeb0/tensorflow/compiler/tests/cast_ops_test.py#L35-L47
 
 I found that the JIT Compiled XLA Bitcast operation (`compiled_f()`) will get the same result with Intel (LE platform) on s390x (BE platform), so I wonder if it has used separate input and output buffers, otherwise after the execution of `compiled_f(x)`, the data in `x` would be changed.
 
Since `f(x)` which invokes the basic tensor Bitcast operation will give BE specific result, the original test would fail on s390x. That's the reason why I think we should do the buffer swapping in `tensor.cc` to make the two different versions of Bitcast operation to get the same result, which aligns with Intel platform as well.
 
Please let me know what you think about this issue, thanks again!

sagunb commented:
Thanks for your explanation Kun-Lu. Could you please update the documentation at 
- tensorflow/tensorflow/tensorflow/go/op/wrappers.go
- tensorflow/tensorflow/tensorflow/compiler/mlir/tensorflow/ir/tf_generated_ops.td
- tensorflow/tensorflow/tensorflow/core/api_def/base_api/api_def_Bitcast.pbtxt

to indicate the a copy is made on BE machines.

kun-lu20 committed:
Update the documentation reg Bitcast op

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Thanks @sagunb ! 

I've updated the above mentioned documentation. Could you please take a look?

kun-lu20 committed:
Fix BUILD file format as per the suggestion by code_check_changed_files

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 committed:
Fix loop reference issue in BUILD file and memory allocation issue

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
CI checks showed several gpu related test cases will fail due to the invocation of `tensor::DeepCopy()`, so I moved this function again into the `else` block to ensure the copy only happens on BE machines.

Also removed the entry `"//tensorflow/core:framework",` from `tensorflow/core/framework/BUILD` to fix the `cycle in dependency graph` issue.

Hi @sagunb , could you please take a look again? Thank you so much!

sagunb commented on tensorflow/core/framework/tensor.cc:
```diff
@@ -755,15 +756,18 @@ Status Tensor::BitcastFrom(const Tensor& other, DataType dtype,
-    Tensor ts_ = tensor::DeepCopy(other);
     if (port::kLittleEndian) {
       buf_ = other.buf_;
+      RefIfNonNull(buf_);
```
did the RefIfNonNull need to be called in the if/else scopes? Since they are always being called - can we leave them outside the conditional as before?

kun-lu20 replied:
Hi @sagunb ,

Yes. I've tried to move `RefIfNonNull(buf_)` out of the conditional block before. However, it didn't work due to the object scope of Tensor instance `ts_`. If `RefIfNonNull(buf_)` isn't in the same scope as `ts_`, tensor data couldn't be retained in `buf_` correctly, so I have to call `RefIfNonNull(buf_)` in both if/else scopes.

kun-lu20 commented on tensorflow/c/c_api_test.cc:
```diff
@@ -1534,11 +1534,13 @@ TEST(CAPI, TestBitcastFrom_Reshape) {
-  EXPECT_EQ(4, *(static_cast<int64_t*>(TF_TensorData(b))));
-  *(static_cast<int64_t*>(TF_TensorData(b))) = 6;
-  EXPECT_EQ(6, *(static_cast<int64_t*>(TF_TensorData(a))));
+  if (tensorflow::port::kLittleEndian) {
```
Hi @cantonios ,

Thanks for the comment. The reason why it does not apply on BE machines is that the output Tensor of Bitcast op does not share a common buffer with the input Tensor, so the write to one tensor does not show up in the other.

cantonios replied:
I understand.  See the other comment.  I think this is a behavior we likely want to preserve.  Especially in this test, where the two bit-casted types are both uint64_t.

cantonios replied:
Well, I'm thinking we likely don't want a deep copy at all, and instead fail if you try to bit-cast something weird like 10 floats to 20 bfloat16s.

std::bit_cast only makes sense when bit-casting between types of the same size.  The only other use-case I'm aware of is for inspecting individual bytes, by bit-casting to a byte tensor (e.g. uint8_t).  *Maybe* there are other valid use-cases, but I don't know of them, and they're probably doing something fishy.  It would be good to know if there are any such cases currently in TF.

kun-lu20 replied:
@cantonios Thanks for the explanation! I think test cases which this PR aims to fix are doing bit-casting between types of different sizes.

For example, `//tensorflow/compiler/tests:cast_ops_test_cpu` is bit-casting from `float16` to `float32`. If we do not apply byte swapping on BE machines, these test cases will fail. And if we use byte swapping without a deep copy, the contents of the input Tensor would be changed after swapping.

kun-lu20 replied:
Hi @cantonios , I still have one question. In test case `//tensorflow/compiler/tests:cast_ops_test_cpu`, I think it's comparing the results between the JIT Compiled XLA Bitcast operation with the basic tensor Bitcast operation. 

https://github.com/tensorflow/tensorflow/blob/6cd264ba65e6161164d7f1b463b7d2415ac0aeb0/tensorflow/compiler/tests/cast_ops_test.py#L35-L47
 
The JIT Compiled XLA Bitcast operation (`compiled_f()`) will give the result in LE format on BE machines, so I guess it has used a copy between input and output buffers, otherwise after the execution of `compiled_f(x)` which involves byte swapping, the data in `x` would've been changed. But through debugging I found the data in `x` wasn't changed. 

If we do not apply deep copy and byte swapping in the basic tensor Bitcast operation, `f(x)` will give BE specific result, and the test will fail on s390x.  

Should we keep the behaviours of different Bitcast implementations consistent on BE machines? Please let me know what you think. Thanks again!

kun-lu20 commented on tensorflow/core/framework/tensor.cc:
```diff
@@ -754,8 +756,18 @@ Status Tensor::BitcastFrom(const Tensor& other, DataType dtype,
+      buf_ = other.buf_;
+      RefIfNonNull(buf_);
+    } else {
+      Tensor ts_ = tensor::DeepCopy(other);
```
Yes, I agree that the first case (bitcasting between types of the same size) does not need a deep copy on BE machines. However, I think the second case (bitcasting to/from (u)int8_t) may still need a deep copy when the type of the other end (source or destination) is not (u)int8_t. Because on BE machines, in general we need two rounds of byte swapping, one for the input type, the other for the output type. 

Please let me know what you think. Thanks @cantonios !

cantonios replied:
Why would you need a deep copy if one of the types is a byte?

You would probably only do this in order to inspect (or potentially modify) individual bytes from the tensor.  Do we need to guarantee that the bytes stream itself would be little endian?  If we don't, it's faster and more consistent to re-use the buffer, with the caller needing to know that bytes are stored in native format.

kun-lu20 replied:
@cantonios Yes, I agree. But I think if we do not keep the data stream as little endian, test cases such as `//tensorflow/compiler/tests:cast_ops_test_cpu` which do the comparison might fail.

cantonios replied:
Did you verify this?  I would still like to avoid byte-swapping when bit-casting to char, assuming a native byte ordering.

kun-lu20 replied:
Thanks @cantonios , I got your point, but If we apply the code changes in this PR to make the behaviours of different Bitcast implementations consistent on BE machines, the above mentioned 3 test cases (`//tensorflow/compiler/tests:cast_ops_test_cpu`, `//tensorflow/compiler/tests:cast_ops_test_cpu_mlir_bridge_test` and `//tensorflow/python/ops/ragged:ragged_bitcast_op_test`) could pass. I've verified this.

Also I've updated the code change to make the copy and swapping only happen on different type sizes on BE machines and reverted the code change in `tensorflow/c/c_api_test.cc`. Could you please take a look? Thank you very much!

kun-lu20 committed:
Add a condition to make the copy only happen on different type sizes

1. Make the copy only happen on different type sizes
2. Revert the test case in c_api_test.cc

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

cantonios reviewed (changes requested):
Can you update the PR description to reflect the changes?

kun-lu20 commented:
Hi @cantonios ,

I've updated the PR description accordingly. Could you please take a look? Thank you!

cantonios reviewed (approved):
Great, thanks.

kun-lu20 committed:
Merge branch 'master' into fix_several_bitcast_related_TCs_on_s390x

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 committed:
Reduce dependency complexity in BUILD file

kun-lu20 commented:
Hi @cantonios ,

I've updated the code to reduce the dependency complexity in `tensorflow/core/framework/BUILD`. Now the `ByteSwapArray()` function only depends on `//tensorflow/core/util/tensor_bundle:byteswaparray` target, which would not cause cyclic dependency. 

Please take a look when you have some time. Thank you very much!

cantonios commented:
We're still seeing internal build failures related to visibility of `tensor_util.h`, specifically for android builds and tf-lite.  I'll need to investigate later today.

cantonios commented:
It is because `tensor_util.h` and its dependencies are not in `framework_internal_private_hdrs`.  We can add `tensor_util.h` in there, but now that depends on `tensor_bundle/byte_swap_array.h` and the like, which are in a different package.  This is causing a layering issue.  We shouldn't have `tensorflow/core/framework:tensor` depend on `tensor_bundle` headers.

The byte swap headers might need to be moved around, and need to be included in `framework_internal_private_hdrs`.  Maybe move them to `util` and add them there?

kun-lu20 commented:
Hi @cantonios , I have a few questions:

1. I checked and found that `tensor_util.h` is already in `//tensorflow/core/framework:framework_internal_private_hdrs`:
https://github.com/tensorflow/tensorflow/blob/1d49a934a6e5bab4e89be1d6d54ddff47834b6eb/tensorflow/core/framework/BUILD#L244

    Do we still need to adjust it?

2. Reg the byte swap headers, do you mean that we should move `byte_swap_array.h` and  `byte_swap_array.cc` into `tensorflow/core/util` folder and add `byte_swap_array.h` to `//tensorflow/core/util:framework_internal_private_hdrs` ?

Thank you very much!

kun-lu20 commented:
Sure, I'll refactor the code and get back to you once it's ready.

Thank you @cantonios !

kun-lu20 committed:
Simplify the dependency layering

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Hi @cantonios , Thanks for your detailed explanation. I've made the following changes:

1. Added `tensor_util.h` to `mobile_srcs_no_runtime` in `tensorflow/core/framework/BUILD`.
2. Moved `byte_swap_array.h` and `byte_swap_array.cc` into `tensorflow/core/util` folder, added these two files to `mobile_srcs_no_runtime` and added  `byte_swap_array.h` to `framework_internal_private_hdrs` in `tensorflow/core/util/BUILD`.
3. Updated other code or BUILD files which reference `byte_swap_array.h` or `byteswaparray` target.

Please take a look when you have some time. Thanks again!

kun-lu20 commented:
Hi @cantonios ,

Hope all is well.

Could you please help me check the status of this PR? Are there any adjustments I should make on my side before it could be merged?

Thank you very much!

cantonios commented:
@kun-lu20 sorry, I've been away for a week.

We still have header issues.  Now it's complaining about
```
./tensorflow/core/framework/tensor_util.h:24:10: error: module //tensorflow/core/framework:tensor does not depend on a module exporting 'tensorflow/core/framework/tensor_shape.proto.h'
```
I'm not sure why we're getting this now, since AFAIK we always included that header.  Maybe we actually rely on it now though with your additions.

kun-lu20 commented:
> @kun-lu20 sorry, I've been away for a week.
> 
> We still have header issues. Now it's complaining about
> 
> ```
> ./tensorflow/core/framework/tensor_util.h:24:10: error: module //tensorflow/core/framework:tensor does not depend on a module exporting 'tensorflow/core/framework/tensor_shape.proto.h'
> ```
> 
> I'm not sure why we're getting this now, since AFAIK we always included that header. Maybe we actually rely on it now though with your additions.

Hi @cantonios ,

No problem. Thanks for the response. I'm not sure how to solve this issue, should I add `":tensor_shape_proto_cc",` in the `deps` of tf_cuda_library `"tensor"` here? https://github.com/tensorflow/tensorflow/blob/c755d039138e6a248170b61d4454096ff219c339/tensorflow/core/framework/BUILD#L788-L827

Thank you!

kun-lu20 commented on tensorflow/compiler/mlir/tensorflow/ir/tf_generated_ops.td:
```diff
@@ -1753,7 +1753,9 @@ tf.Tensor(
 
 *NOTE*: Bitcast is implemented as a low-level cast, so machines with different
-endian orderings will give different results.
+endian orderings will give different results. A copy from input buffer to output 
```
Sure, will do

kun-lu20 commented on tensorflow/core/util/tensor_bundle/byte_swap_tensor.h:
```diff
@@ -19,7 +19,7 @@ limitations under the License.
 #include "tensorflow/core/framework/tensor.h"
```
Yes, I'll move it to `util` folder as well.

cantonios commented:
> should I add `":tensor_shape_proto_cc",` in the `deps` of tf_cuda_library `"tensor"` here?

Yes, that should fix it, thanks.

kun-lu20 commented:
Hi @cantonios ,

I just checked the master branch and found that this commit https://github.com/tensorflow/tensorflow/commit/2d3eb4071b63ae55cad1c198b8292d90ae761b87 has already moved core/util/tensor_bundle/byte_swap_array.{cc,h} to TSL and it didn't move byte_swap_tensor.h with them. 

Should we follow it and keep byte_swap_array.{cc,h} in TSL? Thank you!

cantonios commented:
> Should we follow it and keep byte_swap_array.{cc,h} in TSL? Thank you!

Yes, everything should be kept together, thanks.

kun-lu20 commented:
OK, I'll follow the practice in master branch. Thank you.

kun-lu20 committed:
Merge branch 'master' into fix_several_bitcast_test_1017

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 committed:
Fix the header issue and remove trailing spaces

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Hi @cantonios ,

Thanks for your advice. The code changes of this round include:

1. Fix the header issue by adding `:tensor_shape_proto_cc` to `deps`.
2. Remove the trailing spaces from the description lines.
3. Use the way consistent with master branch to deal with `byte_swap_array` dependency.

Please review the code change when you have some time. Thanks again!

cantonios commented on tensorflow/core/framework/tensor.cc:
```diff
@@ -58,6 +59,7 @@ limitations under the License.
 #include "tensorflow/core/platform/protobuf.h"
 #include "tensorflow/core/platform/tensor_coding.h"
 #include "tensorflow/core/platform/types.h"
+#include "tensorflow/tsl/util/tensor_bundle/byte_swap_array.h"
```
Sorry I didn't notice in your last comment, but these should be directly under `util`, and not `tensor_bundle` anymore.  Otherwise we're going to run into the same layering issue as before.

kun-lu20 replied:
Got it. I'll move `byte_swap_array.h` and `byte_swap_array.cc` to `tsl/util` folder.

cantonios commented on tensorflow/tsl/util/tensor_bundle/BUILD:
```diff
@@ -22,10 +22,25 @@ cc_library(
+filegroup(
+    name = "framework_internal_private_hdrs",
+    srcs = [
+        "byte_swap_array.h",
```
You may as well as the `byte_swap_tensor.h` utility here too, for consistency (but again, moved under tsl/util).

kun-lu20 replied:
Sure, I'll move them to `tsl/util` folder as well. Thanks!

cantonios replied:
okay, sounds good

kun-lu20 committed:
Move byte_swap_array target to tsl/util

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 committed:
Merge branch 'master' into fix_several_bitcast_test_1018

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Hi @cantonios ,

I've moved `byte_swap_array.h` and `byte_swap_array.cc` to `tsl/util`, please take a look when you have some time. Thank you!

kun-lu20 commented:
Hi @cantonios ,

Could you please review the code change when you have some time? Thank you very much!

kun-lu20 committed:
Fix the target missing error in tensor_bundle BUILD file

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Hi @cantonios ,

I just checked the `Code Check - Full` result and realized that I forgot to update the `//tensorflow/tsl/util/tensor_bundle:mobile_srcs` target in `tensorflow/core/util/tensor_bundle/BUILD`.

Since the `tensor_bundle` folder under `tsl/util` was removed, I added a file group in `tensorflow/tsl/util/BUILD` and used it to replace the missing target in `tensorflow/core/util/tensor_bundle/BUILD`.

Please review the latest code change when you have some time. Thanks very much for your time and effort!

kun-lu20 commented:
Hi @cantonios ,

Hope all is well.

Could you please check this PR when you have some time? Are there any other header issues which need to be addressed? 

Thank you!

cantonios commented:
> Could you please check this PR when you have some time? Are there any other header issues which need to be addressed?

Yes, we're still seeing errors of the form:
```
tensorflow/core/framework/tensor.cc:63:10: fatal error: 'tensorflow/tsl/util/byte_swap_array.h' file not found
```
for mobile builds.  Once all the tests pass, this should be submitted automatically.

cantonios commented on tensorflow/core/util/tensor_bundle/BUILD:
```diff
@@ -27,7 +27,7 @@ filegroup(
         "tensor_bundle.cc",
         "tensor_bundle.h",
-        "//tensorflow/tsl/util/tensor_bundle:mobile_srcs",
+        "//tensorflow/tsl/util:mobile_srcs_byteswaparray",
```
this isn't needed, remove (it will already be included by tsl/util).

kun-lu20 replied:
Removed

cantonios commented on tensorflow/tsl/util/BUILD:
```diff
@@ -26,9 +26,19 @@ package(
 )
 
+filegroup(
+    name = "mobile_srcs_byteswaparray",
```
Rename this to "mobile_srcs_no_runtime".

In `//tensorflow/core/util/BUILD`, under "mobile_srcs_no_runtime" add this to "srcs":
```
"//tensorflow/tsl/util:mobile_srcs_no_runtime",
```
Turns out this is the missing dependency for the broken android builds.

kun-lu20 replied:
Done. Thanks!

cantonios commented on tensorflow/tsl/util/BUILD:
```diff
@@ -26,9 +26,19 @@ package(
 filegroup(
     name = "mobile_srcs_only_runtime",
     srcs = [
+        "byte_swap_array.cc",
+        "byte_swap_array.h",
```
Remove these from here since they are needed for `mobile_srcs_no_runtime`, and including them in both leads to duplicate symbol errors.

kun-lu20 replied:
Removed.

cantonios reviewed (changes requested):
My added comments address at least some of the mobile build errors.  I'm running through CI again to make sure the requested changes don't break anything else.

kun-lu20 committed:
Fix the mobile build errors

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

cantonios reviewed (approved):
Thanks, 42nd try the charm, right?  Thanks for your patience.  Hopefully this will go through now.

kun-lu20 commented:
Yes, exactly. 

Hope so. Thank you so much for your help @cantonios !

cantonios commented:
🎉 🎉

kun-lu20 commented:
@cantonios Thank you so much 😀
